### PR TITLE
Update num_tokens_across_dp to use nccl instead of gloo

### PR DESCRIPTION
**Pull request by SageMoore**
## Purpose
This PR changes the All Reduce backend that we use to synchronize DP padding between ranks from gloo to nccl. This AR starts to be a significant source of overhead when running with multi-node setups. Anecdotally, we've seen the Gloo backed all reduce take ~10ms in a two node cluster setup. Simply swapping over to NCCL brings the over head down to 100s of microseconds. 

## Test Plan
lm_eval with the following serving command seems sufficient: `vllm serve --model="deepseek-ai/DeepSeek-V2-Lite" --data-parallel-size 2 --enable-expert-parallel `

I'm happy to run additional models/setups if we think that's necessary.
## Test Result
```
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.3700|±  |0.0279|
|     |       |strict-match    |     5|exact_match|↑  |0.3633|±  |0.0278|
```

### CODE REVIEW

File: vllm/forward_context.py [hunk 1]
```diff
@@ -78,10 +78,10 @@ def num_tokens_across_dp(num_tokens: int, dp_size: int,
         num_tokens_across_dp = [0] * dp_size
         num_tokens_across_dp[dp_rank] = num_tokens
         num_tokens_tensor = torch.tensor(num_tokens_across_dp,
-                                         device="cpu",
+                                         device="cuda",
```
Comment (tlrmchlsmth) [inline vllm/forward_context.py]: I think it needs to be:
```suggestion
                                         device=current_platform.device_type,
```

File: vllm/forward_context.py [hunk 2]
```diff
@@ -78,10 +78,10 @@ def num_tokens_across_dp(num_tokens: int, dp_size: int,
...
                                          dtype=torch.int32)
         from vllm.distributed.parallel_state import get_dp_group
-        dist.all_reduce(num_tokens_tensor, group=get_dp_group().cpu_group)
+        dist.all_reduce(num_tokens_tensor, group=get_dp_group().device_group)
```
Comment (tlrmchlsmth) [inline vllm/forward_context.py]: and then you'll need to copy it from the device -> CPU as well

Comment (LucasWilkinson) [reply vllm/forward_context.py]: can we use an env var to switch between a CPU all-reduce and device one for now? (we can default it to device); its just the device->CPU transfer will cause a sync point and break the overlap in https://github.com/vllm-project/vllm/pull/23569 so it would be nice to have a knob for use to experiment with (since with the overlap the CPU all-reduce will likely be overlapped too)

**Comment by SageMoore** [comment]
Here are some profiles for a DP=4 single node run.

With Gloo
Screenshot 2025-08-29 at 4 03 21 PM

With NCCL
Screenshot 2025-08-29 at 4 04 13 PM

File: vllm/envs.py [hunk 3]
```diff
@@ -739,6 +740,10 @@ def get_vllm_port() -> Optional[int]:
     lambda: [] if "VLLM_DISABLED_KERNELS" not in os.environ else os.environ[
         "VLLM_DISABLED_KERNELS"].split(","),
 
+    "VLLM_DISABLE_NCCL_DP_PADDING":
+    lambda: (os.getenv("VLLM_DISABLE_NCCL_DP_PADDING", "False").lower() in
+             ("true", "1")),
```
Comment (tlrmchlsmth) [inline vllm/envs.py]: nit: I find this variable name a bit confusing - the way I read it, it sounds like there is some sort of padding that we're disabling. Maybe rename it to `VLLM_USE_NCCL_FOR_DP_SYNCHRONIZATION`?

Comment (tlrmchlsmth) [reply vllm/envs.py]: Could you add a comment as well? (otherwise PR looks good)

Comment (SageMoore) [reply vllm/envs.py]: How about `VLLM_DISABLE_NCCL_FOR_DP_SYNCHRONIZATION`?

**Review by tlrmchlsmth — APPROVED** [review]
